Add unit and e2e tests to release workflow (#202)
kyma-project/btp-manager: docs/workflows.md

Both scripts are run from the workflow but can also be triggered manually from the developer's computer. They are placed under `hack/update/`.

To keep `module-chart/chart` in sync with the [upstream](https://github.com/SAP/sap-btp-service-operator), you must not apply any manual changes there.

## Release pipeline

See [BTP Manager release pipeline](release.md)

## Run E2E tests

> Triggered by pull requests on `main` branch calls reusable workflow 'Run E2E tests (reusable)' (`run-e2e-tests-reusable.yaml`). 

Reviewer: I need more information here.

Author: Corrected in the next commit.

> This workflow is triggered by pull requests on `main` branch. Uses DEV artifact registry, tags binary and OCI module image with PR-number and calls reusable workflow 

Reviewer: ```suggestion
This workflow is triggered by pull requests on the `main` branch. It uses the DEV artifact registry, tags the binary and OCI module image with the PR number, and calls the reusable workflow 
```

Uses DEV artifact registry and tags binary and image with PR-number. 

## Run unit tests

> Triggered by pull requests on `main` branch and for releases (tag creation on 'main' branch) calls reusable workflow 'Run unit tests (reusable)' (`run-unit-tests-reusable.yaml`).

Author: Corrected in the next commit.

> This workflow is triggered by pull requests on `main` branch. Then calls reusable workflow 'Run unit tests (reusable)' (`run-unit-tests-reusable.yaml`).

Reviewer: ```suggestion
This workflow is triggered by pull requests on the `main` branch. Then it calls the reusable workflow 'Run unit tests (reusable)' (`run-unit-tests-reusable.yaml`).
```

## Reusable workflows

There are reusable workflows created. Anyone with access to the reusable workflow can call the reusable workflow from another workflow.

### Run E2E tests

This workflow runs the E2E tests on the k3a cluster. 
From the calling workflow you pass the following parameters:

| Parameter name  | Required | Description |
| ------------- | ------------- | ------------- |

> | image-repo  | yes  | Binary image registry reference |
> | module-repo  | yes  |  OCI module image registry reference |
> | image-tag  | yes  |  Binary image tag |
> | module-tag  | yes  |  OCI module image tag |
> | skip-templates  | no  |  wait for images only, skip other artifacts |

Reviewer: ```suggestion
| **image-repo**  | yes  | binary image registry reference |
| **module-repo**  | yes  |  OCI module image registry reference |
| **image-tag**  | yes  |  binary image tag |
| **module-tag**  | yes  |  OCI module image tag |
| **skip-templates**  | no  |  wait for images only, skip other artifacts |
```

Pipeline:

> - prepares k3s cluster and the docker registry,
> - waits for the artifacts to be ready in the registry,
> - runs the E2E tests on the cluster.

Reviewer: ```suggestion
- prepares a k3s cluster and the Docker registry
- waits for the artifacts to be ready in the registry
- runs the E2E tests on the cluster
```

### Run unit tests

This workflow runs the unit tests.
No parameters are passed from calling pipeline (callee).

> Pipeline:

Reviewer: ```suggestion
The pipeline:
```

> - checkouts code and sets up the cache,
> - sets up go environment,

Reviewer: ```suggestion
- checks out code and sets up the cache
- sets up the Go environment
```

> - invokes `make test`.

Reviewer: ```suggestion
- invokes `make test`
```
